only clean query parameters

cebtenzzre opened the pull request:
There is no need to "clean" the json body of a POST request to alter things like booleans - JSON already has a way to represent them. That wasn't done in 4.4.x, which had a more transparent set_state.

One possible exception is datetime, which doesn't have one true JSON representation. But we can enhance convert_json to always use isoformat.

(Also, DELETE shouldn't have a body - but it may have query params. We don't currently use kwargs with DELETE so this isn't crucial, but it's important for correctness.)

Fixes #2464

cebtenzzre committed:
only clean query parameters

jsl12 committed:
using urlib for making http endpoints, fixes delete endpoint

jsl12 committed:
added to history

cebtenzzre committed:
fix HASS url building

jsl12 committed:
Merge branch 'dev' into pr/cebtenzzre/2475

jsl12 reviewed (approved):
Using urllib to actually parse things should resolve this stuff about where the slashes are, so I don't think #2485 is necessary any more

cebtenzzre committed:
use yarl.URL and handle automatic trailing slash

cebtenzzre commented on appdaemon/utils.py:
```diff
@@ -1167,13 +1173,11 @@ def clean_http_kwargs(val: Any) -> Any:
-    return result.strip("/")
+    parsed = parse.urlparse(str(base))
+    result = parsed._replace(path=endpoint)
+    return parse.urlunparse(result)
```
I think this breaks base URLs with a path (e.g. `http://supervisor/core`) - it will replace `/core` with the endpoint path instead of appending. See #2485 for a better fix using yarl.

cebtenzzre committed:
fix PydanticSchemaGenerationError

We need a serializer for all fields in the model.

cebtenzzre committed:
just use a property

cebtenzzre committed:
avoid urljoin

We prefix paths like `/api/states` with a slash even though they're
relative in practice, not absolute. It's simpler to always append to the
path and lstrip the leading slash from the endpoint.

cebtenzzre committed:
simplify

jsl12 committed:
keeping the native yarl.URL object with arbitrary_types_allowed

jsl12 committed:
Merge branch 'pr/cebtenzzre/2485' into pr/cebtenzzre/2475

jsl12 committed:
pruned make_endpoint

jsl12 committed:
pruned unused imports
